### PR TITLE
Refactor wakunode2.nim

### PR DESCRIPTION
**Pull request by jm-clius**
## Background

This PR focuses on refactoring `wakunode2.nim` according to the [Track 1 refactoring checklist](https://hackmd.io/1imOGULZRsed2HpgmzGleA?view)

As a result multiple referenced modules required similar refactoring and improvements, especially related to error handling.

The first step towards improved error handling is being explicit about what exceptions we raise, and to avoid generic types such as `Exception` or `CatchableError`. A next step is to determine which of these raised exceptions we should catch, though such changes fall outside the scope of this single PR.

## Note:
- Some underlying modules, especially utils, still raise `Exception` where they should more properly return a `Result`. This will be addressed in follow-up PRs, once those modules undergo Track 1 refactoring.
- `import`s are likely to be cleaned up further once referenced modules export their own referenced types.
- `waitFor` raises `CatchableError`. In future we should probably catch these and raise our own error.
- `metrics` raises `Exception` in some places. Could be part of our upstream effort to fix this/log issue?

### CODE REVIEW

File: waku/v2/node/peer_manager/peer_manager.nim [hunk 1]
```diff
@@ -162,7 +162,7 @@ proc hasPeers*(pm: PeerManager, proto: string): bool =
   # Returns `true` if manager has any peers for the specified protocol
   pm.peers.anyIt(it.protos.contains(proto))
 
-proc addPeer*(pm: PeerManager, peerInfo: PeerInfo, proto: string) {.raises: [Defect, Exception]} =
```
Comment (staheri14) [inline waku/v2/node/peer_manager/peer_manager.nim:165]: Could you please explain why the `raises` pragma is removed in here?

Comment (staheri14) [reply waku/v2/node/peer_manager/peer_manager.nim:165]: I think that is because of the `{.push raises: [Defect].}` https://github.com/status-im/nim-waku/blob/37c5a6eb222921375a2f9bcedc9133b2c4b1b705/waku/v2/node/peer_manager/peer_manager.nim#L1` at the top of the file

File: waku/v2/node/storage/peer/waku_peer_storage.nim [hunk 2]
```diff
@@ -47,9 +49,12 @@ proc encode*(storedInfo: StoredInfo): ProtoBuffer =
   for proto in storedInfo.protos.items:
     pb.write(3, proto)
   
-  pb.write(4, storedInfo.publicKey)
+  try:
+    pb.write(4, storedInfo.publicKey)
+  except ResultError[CryptoError] as e:
```
Comment (staheri14) [inline waku/v2/node/storage/peer/waku_peer_storage.nim:54]: I am asking this for my own understanding, I thought we should use `Result` when multiple failure paths exist, is it the case here? It seems there is only one failure path

File: waku/v2/node/storage/peer/waku_peer_storage.nim [hunk 3]
```diff
@@ -107,7 +117,7 @@ method getAll*(db: WakuPeerStorage, onData: peer_storage.DataProc): PeerStorageR
   ## Retrieves all peers from storage
   var gotPeers = false
 
-  proc peer(s: ptr sqlite3_stmt) = 
+  proc peer(s: ptr sqlite3_stmt) {.raises: [Defect, LPError, ResultError[ProtoError]].} =
```
Comment (staheri14) [inline waku/v2/node/storage/peer/waku_peer_storage.nim:120]: Should we use `{.raises.}` for non public procs?

File: waku/v2/node/wakunode2.nim [hunk 4]
```diff
@@ -406,14 +395,14 @@ proc mountFilter*(node: WakuNode) =
 
 # NOTE: If using the swap protocol, it must be mounted before store. This is
 # because store is using a reference to the swap protocol.
-proc mountSwap*(node: WakuNode, swapConfig: SwapConfig = SwapConfig.init()) =
+proc mountSwap*(node: WakuNode, swapConfig: SwapConfig = SwapConfig.init()) {.raises: [Defect, LPError].} =
```
Comment (staheri14) [inline waku/v2/node/wakunode2.nim:398]: There is already a `{.push raises: [Defect].}` at the top, do we need another one in here?

File: waku/v2/node/wakunode2.nim [hunk 5]
```diff
@@ -406,14 +395,14 @@ proc mountFilter*(node: WakuNode) =
...
   info "mounting swap", mode = $swapConfig.mode
   node.wakuSwap = WakuSwap.init(node.peerManager, node.rng, swapConfig)
   node.switch.mount(node.wakuSwap, protocolMatcher(WakuSwapCodec))
   # NYI - Do we need this?
   #node.subscriptions.subscribe(WakuSwapCodec, node.wakuSwap.subscription())
 
-proc mountStore*(node: WakuNode, store: MessageStore = nil, persistMessages: bool = false) =
+proc mountStore*(node: WakuNode, store: MessageStore = nil, persistMessages: bool = false) {.raises: [Defect, LPError].} =
```
Comment (staheri14) [inline waku/v2/node/wakunode2.nim:405]: Ditto

File: waku/v2/node/wakunode2.nim [hunk 6]
```diff
@@ -529,7 +522,7 @@ proc mountRelay*(node: WakuNode,
 
     info "relay mounted and started successfully"
 
-proc mountLightPush*(node: WakuNode) =
+proc mountLightPush*(node: WakuNode) {.raises: [Defect, LPError].} =
```
Comment (staheri14) [inline waku/v2/node/wakunode2.nim:525]: Ditto

File: waku/v2/node/wakunode2.nim [hunk 7]
```diff
@@ -541,10 +534,15 @@ proc mountLightPush*(node: WakuNode) =
   
   node.switch.mount(node.wakuLightPush, protocolMatcher(WakuLightPushCodec))
 
-proc mountLibp2pPing*(node: WakuNode) =
+proc mountLibp2pPing*(node: WakuNode) {.raises: [Defect, LPError].} =
```
Comment (staheri14) [inline waku/v2/node/wakunode2.nim:537]: Ditto

File: waku/v2/protocol/waku_store/waku_store.nim [hunk 8]
```diff
@@ -396,7 +396,7 @@ proc init*(ws: WakuStore) {.raises: [Defect, Exception]} =
 
 
 proc init*(T: type WakuStore, peerManager: PeerManager, rng: ref BrHmacDrbgContext,
-                   store: MessageStore = nil, wakuSwap: WakuSwap = nil, persistMessages = true): T {.raises: [Defect, Exception]} =
```
Comment (staheri14) [inline waku/v2/protocol/waku_store/waku_store.nim:399]: Why is this removed?

**Review by staheri14 — COMMENTED** [review]
Looks good to me! Just had a few questions :)

File: waku/v2/node/storage/peer/waku_peer_storage.nim (hunk 2 above)
Comment (staheri14) [reply waku/v2/node/storage/peer/waku_peer_storage.nim:54]: I have a general question regarding the use of `Result` but related to the files in this PR, if you check line https://github.com/status-im/nim-waku/blob/37c5a6eb222921375a2f9bcedc9133b2c4b1b705/waku/v2/node/storage/peer/waku_peer_storage.nim#L23, the output is defined as `ProtoResult[T]` however, this proc never results in an error, why do we need the Result type? I am going over the waku protocol modules and was wondering whether we should remove such `Result` usage or not?

File: waku/v2/node/storage/peer/waku_peer_storage.nim (hunk 3 above)
Comment (oskarth) [reply waku/v2/node/storage/peer/waku_peer_storage.nim:120]: I think it isn't a must but in general it is prudent as it helps us. Public functions more about providing clean interface to consumer with clear error paths. At least that's how I understand.

File: waku/v2/node/wakunode2.nim [hunk 9]
```diff
@@ -155,7 +144,7 @@ proc init*(T: type WakuNode, nodeKey: crypto.PrivateKey,
   #    msgIdProvider = msgIdProvider,
   #    triggerSelf = true, sign = false,
   #    verifySignature = false).PubSub
-  result = WakuNode(
+  return WakuNode(
```
Comment (oskarth) [inline waku/v2/node/wakunode2.nim:147]: Nitpick but I'd probably first init it then `return wakuNode`. Just because we are creating a new peer manager etc, less stuff going on in one expression/statement

File: waku/v2/node/wakunode2.nim [hunk 10]
```diff
@@ -393,9 +382,9 @@ proc info*(node: WakuNode): WakuInfo =
   let wakuInfo = WakuInfo(listenStr: listenStr)
   return wakuInfo
 
-proc mountFilter*(node: WakuNode) =
+proc mountFilter*(node: WakuNode) {.raises: [Defect, KeyError, LPError]} =
   info "mounting filter"
-  proc filterHandler(requestId: string, msg: MessagePush) {.gcsafe.} =
+  proc filterHandler(requestId: string, msg: MessagePush) {.gcsafe, raises: [Defect, KeyError].} =
```
Comment (oskarth) [inline waku/v2/node/wakunode2.nim:387]: Nitpick but if line is too long might want to pull down whole pragma on newline

File: waku/v2/node/wakunode2.nim [hunk 11]
```diff
@@ -478,7 +470,8 @@ proc mountRelay*(node: WakuNode,
                  topics: seq[string] = newSeq[string](),
                  rlnRelayEnabled = false,
                  relayMessages = true,
-                 triggerSelf = true) {.gcsafe.} =
+                 triggerSelf = true)
+                 {.gcsafe, raises: [Defect, InitializationError, LPError, CatchableError].} = # @TODO: Better error handling: CatchableError is raised by `waitFor`
```
Comment (oskarth) [inline waku/v2/node/wakunode2.nim:474]: This form is a bit hard long and hard to read, is it possible to format at different somehow?

**Review by oskarth — APPROVED** [review]
Thanks!

File: waku/v2/node/peer_manager/peer_manager.nim (hunk 1 above)
Comment (jm-clius) [reply waku/v2/node/peer_manager/peer_manager.nim:165]: True, but also because the underlying exception is now caught and handled elsewhere in the `peer_manager` and therefore not raised by `addPeer`.

File: waku/v2/node/storage/peer/waku_peer_storage.nim (hunk 2 above)
Comment (jm-clius) [reply waku/v2/node/storage/peer/waku_peer_storage.nim:54]: Yeah, I don't have strong feelings about this, but I think `Result` is also prudent to use where we can imagine multiple failure paths in future. Encoding/decoding is a good example, because this can often fail in multiple ways. I'd rather elevate more return types to `Result` than remove `Result` for functions where we don't return `err` yet. There's also the consistency of other protobuf `init*` functions returning `ProtoResult`, and we probably want to maintain this pattern. Another reason to use `Result` above `Opt`, for example, is that you can be explicit about what caused the error.

File: waku/v2/node/storage/peer/waku_peer_storage.nim (hunk 3 above)
Comment (jm-clius) [reply waku/v2/node/storage/peer/waku_peer_storage.nim:120]: In some cases, where non public procs are used by public procs, it becomes necessary to add explicit `{.raises.}`.
In this instance, for example, we can be explicit about what `peer()` raises and then catch and handle the exceptions [here](https://github.com/status-im/nim-waku/blob/37c5a6eb222921375a2f9bcedc9133b2c4b1b705/waku/v2/node/storage/peer/waku_peer_storage.nim#L139-L142). This leaves the public proc free of exceptions.

File: waku/v2/node/wakunode2.nim (hunk 4 above)
Comment (jm-clius) [reply waku/v2/node/wakunode2.nim:398]: The top-level gets _replaced_ by the function specific raises `{.raises: [Defect, LPError].}`. https://status-im.github.io/nim-style-guide/errors.exceptions.html. `{.raises: [Defect]}` is too strict for this function which raises an `LPError` too, hence it needs its own annotation.

File: waku/v2/node/wakunode2.nim (hunk 5 above)
Comment (jm-clius) [reply waku/v2/node/wakunode2.nim:405]: See above

File: waku/v2/protocol/waku_store/waku_store.nim (hunk 8 above)
Comment (jm-clius) [reply waku/v2/protocol/waku_store/waku_store.nim:399]: In general, raising the non-specific `Exception` is bad practice. Here it was done because the underlying modules weren't explicit about what they're raising. Now that the underlying modules have better error-handling, the top-level `{.push raises: [Defect].}` is enough for this function.

File: waku/v2/node/wakunode2.nim (hunk 11 above)
Comment (jm-clius) [reply waku/v2/node/wakunode2.nim]: I've moved the comment to above the pragma line and the pragma to a single indent inside the proc block [here](https://github.com/status-im/nim-waku/blob/47981cfdf527fed4728462794e5e41d381c2a2b9/waku/v2/node/wakunode2.nim#L478-L479). Let me know if this looks reasonable to you.

File: waku/v2/node/wakunode2.nim (hunk 10 above)
Comment (jm-clius) [reply waku/v2/node/wakunode2.nim]: Done in 47981cf

File: waku/v2/node/wakunode2.nim (hunk 9 above)
Comment (jm-clius) [reply waku/v2/node/wakunode2.nim]: Done. 47981cf

File: waku/v2/node/peer_manager/peer_manager.nim (hunk 1 above)
Comment (staheri14) [reply waku/v2/node/peer_manager/peer_manager.nim:165]: Thanks for the clarification!

File: waku/v2/node/storage/peer/waku_peer_storage.nim (hunk 2 above)
Comment (staheri14) [reply waku/v2/node/storage/peer/waku_peer_storage.nim:54]: Thanks for the reply! 
What do you refer to by `Opt`, is it `Option`?

File: waku/v2/node/storage/peer/waku_peer_storage.nim (hunk 3 above)
Comment (staheri14) [reply waku/v2/node/storage/peer/waku_peer_storage.nim:120]: I see, I could not find any exact item on the track1 checklist for this one, good to know! makes sense, I will consider it while refactoring other modules.

**Review by staheri14 — APPROVED** [review]
Thanks!

File: waku/v2/node/storage/peer/waku_peer_storage.nim (hunk 2 above)
Comment (jm-clius) [reply waku/v2/node/storage/peer/waku_peer_storage.nim:54]: Indeed!
